Improve surface and buffer handling

ikalco opened the pull request:
#### Describe your PR, what does it fix/add?
main points here
* wait on dmabuf buffers
* make renderer correctly lock/unlock buffers
* make CHLBufferReference actually work as intended
* conform more to wayland protocol, specifically for wl_surface 
* cleanup surface and buffer stuff from previous prs to be more readable

#### Is there anything you want to mention? (unchecked code, possible bugs, found problems, breaking compatibility, etc.)
Was going to make direct scanout use CHLBufferReference too, but I think it'll also take aq pr so I'll get around to it later

btw if you want me to split this into multiple PRs, I prob can

#### Is it ready for merging, or does it need work?
yes

vaxerski commented:
either split it to multiple PRs or rebase it nicely so that each commit is like a PR so we can review it and merge linux kernel style

PlasmaPower commented on src/managers/eventLoop/EventLoopManager.cpp:
```diff
@@ -143,6 +187,39 @@ void CEventLoopManager::doLater(const std::function<void()>& fn) {
 }
 
+void CEventLoopManager::doOnAllReadable(const std::vector<int>& fds, const std::function<void()>& fn) {
+    if (fds.empty() || std::ranges::all_of(fds, [](int i) { return i == -1; })) {
```
Shouldn't we filter out any -1 fds instead of checking the case where they're all -1? E.g. what if only one of the fds is -1, wouldn't that throw a wrench into the event loop?

ikalco replied:
yeah might as well just do it at the start

PlasmaPower commented on src/protocols/DRMSyncobj.cpp:
```diff
@@ -75,98 +67,40 @@ CDRMSyncobjSurfaceResource::CDRMSyncobjSurfaceResource(UP<CWpLinuxDrmSyncobjSurf
+        surface->pending.buffer->acquire = makeUnique<CDRMSyncPointState>(std::move(pendingAcquire));
+        pendingAcquire                   = {};
+
+        surface->pending.buffer->release = makeUnique<CDRMSyncPointState>(std::move(pendingRelease));
```
Actually I think the spec is vague on this. I'm not sure if it's supposed to be per-commit, per-surface, or per-buffer. I read it as the first one but I don't think there's as definitive of an answer as I originally thought.

ikalco replied:
Yeah I'm leaning more towards it being kept in surface state because the spec says
> Thus, it applies only to the buffer that is attached to the surface at commit time.

and also it would fix that scenario + there's not really a downside to moving the points from per-buffer to per-surface I don't think

ikalco replied:
actually, it makes more sense to keep track of release points in buffer
and keep/use acquire point in surface state

ikalco commented on src/protocols/core/Compositor.cpp:
```diff
@@ -109,36 +116,83 @@ CWLSurfaceResource::CWLSurfaceResource(SP<CWlSurface> resource_) : resource(reso
+            it->ready = true;
+
+            // to preserve order, only the front buffer can ever be comitted
+            if (pendingStates.front() != it)
```
I thought of that initially too but it makes it look like frames skip
and there is other state that needs to be applied aswell, other than buffer

PlasmaPower replied:
For the non-buffer state you could just trigger all the commits at once here, but if it looks bad then yeah better to just keep it as-is.

ikalco commented on src/managers/eventLoop/EventLoopManager.cpp:
```diff
@@ -50,6 +58,42 @@ static int configWatcherWrite(int fd, uint32_t mask, void* data) {
+    }
+
+    if (mask & WL_EVENT_READABLE) {}
+    g_pEventLoopManager->removeReadableWaiterSource((CEventLoopManager::SReadableWaiterSource*)data);
```
oops, yes lol

ikalco commented:
I'll clean up the pr into multiple nice commits sometime tmr, probably based on those main points

ikalco commented:
also vaxry do you care if I make the SSurfaceState updated thing a struct of bools

would be nice to do `pending.updated.buffer` instead of 
`const bool newBuffer  = pending.updated & SSurfaceState::SURFACE_UPDATED_BUFFER;`

gulafaran commented:
23300cf1d69ab66aa8d639adf8932ab5b7c117c1
 i dont like the idea of removing that entire syncreleaser class and doing that but i do like following protocol.
 
so strip out the commits that makes it `conform more to wayland protocol, specifically for wl_surface` and make a PR with those first. because this rewrites the entire buffer releasing yet again. from past 4 weeks of fiddling with this im sure this will require a whole heap of rounds of testing yet again to not break vrr/tearing/ds and all sorts of hidden shenanigans. even benchmarking.

gulafaran commented:
[2abb6e5](https://github.com/hyprwm/Hyprland/pull/9774/commits/2abb6e5e1a033740e7ba8625ee84e8c7ed712f19) is probably making non syncobj buffers following protocol, but this with the current removal of syncreleaser means that the buffer wont signal the release point it had recieved and will most likely cause clients to keep it referenced forever or even make VK_PRESENT_MODE_FIFO_KHR not work at all.

gulafaran commented:
[dd47b80](https://github.com/hyprwm/Hyprland/pull/9774/commits/dd47b80914036775e0e7f674ec942f95a619626f) is something that probably is wanted but i never proceeded with because it requires some more things than just checking the dmabuf fd itself, because of 

```
Userspace can perform a DMA_BUF_IOCTL_EXPORT_SYNC_FILE to retrieve the current set of fences on a dma-buf file descriptor as
a sync_file. CPU waits via poll() or other driver-specific mechanisms typically wait on whatever fences are on the dma-buf at the 
time the wait begins. This is similar except that it takes a snapshot of the current fences on the dma-buf for waiting later instead 
of waiting immediately. This is useful for modern graphics APIs such as Vulkan which assume an explicit synchronization model 
but still need to inter-operate with dma-buf.
```

something like

```
    dma_buf_export_sync_file request{
        .flags = DMA_BUF_SYNC_READ,
        .fd = -1,
    };
    if (drmIoctl(FD, DMA_BUF_IOCTL_EXPORT_SYNC_FILE, &request) == 0) {
        return CFileDescriptor(request.fd);
    }

```

but that is linux only. https://dri.freedesktop.org/docs/drm/driver-api/dma-buf.html#c.dma_buf_export_sync_file

PlasmaPower commented:
It seems like there's two main behavior changes of this PR:
1. Acquire: This PR moves acquire waiting into the surface code instead of the syncobj, and waits on buffers to become readable to optimize the implicit sync.
2. Release: This PR merges the release event and release syncobj signaling code so that they're signaled at the same time, instead of the current code which signals the release event when we're done with the buffer (implicit sync) and the release syncobj when the GPU is done with the buffer (explicit sync).

The first change makes a lot of sense to me, but I'm not sure the second change is necessary. Unless it seems to reduce implicit sync client flickering on nvidia I'd favor leaving it as-is. I think implicit sync flickering on nvidia is due to the acquire being early not the release.

PlasmaPower commented:
Wow I just checked and even old versions of electron have sync_files attached to the dmabufs it sends us, so I think we might be able to get explicit sync from all sorts of apps that previously only supported implicit sync! We should definitely try that out and see if it fixes e.g. discord flickering on nvidia.

ikalco commented:
> Release: This PR merges the release event and release syncobj signaling code so that they're signaled at the same time, instead of the current code which signals the release event when we're done with the buffer (implicit sync) and the release syncobj when the GPU is done with the buffer (explicit sync).

they should be the same, both implicit and explicit sync should be signaled when we are done with the buffer
and we are only done with the buffer after rendering is done
So, they should be signaled at the same time

PlasmaPower commented:
With implicit sync we're done with the buffer as soon as we glFlush, and the GPU is supposed to implicitly handle synchronization. I know nvidia doesn't implement that, but I'm not sure it causes any issues in practice (compared to the lack of an acquire point which definitely does).

PlasmaPower commented:
I'm not against delaying the release event in theory, but I worry about the complexity of this change and all the edge cases we've run into with previous explicit sync changes, and I'm not sure there's practical benefit to the release changes. Perhaps what we need is an explicit sync test suite to help give us confidence in changes like this.

ikalco commented:
only pushed to keep track of commits
about to split into prs

ikalco commented:
closing now cause other prs are up
